File: dashboard/src/lib/api/bundles.ts
import { API_BASE, ApiError, request } from "./client"

// --- Bundle summary (from GET /bundles) ---

export interface BundleSummary {
  name: string
  latest_version: number
  version_count: number
  last_updated: string
  deployed_envs: string[]
}

// --- Bundle response ---

export interface BundleResponse {
  id: string
  tenant_id: string
  name: string
  version: number
  revision_hash: string
  signature_hex: string | null
  source_hub_slug: string | null
  source_hub_revision: string | null
  uploaded_by: string
  created_at: string
}

export interface BundleWithDeployments extends BundleResponse {
  deployed_envs: string[]
}

export interface DeploymentResponse {
  id: string
  env: string
  bundle_name: string
  bundle_version: number
  deployed_by: string
  created_at: string
}

/** List distinct bundle names with summaries. */
export function listBundles() {
  return request<BundleSummary[]>("/bundles")
}

/** List all versions for a named bundle. */
export function listBundleVersions(name: string) {
  return request<BundleWithDeployments[]>(
    `/bundles/${encodeURIComponent(name)}`,
  )
}

/** Upload a new bundle version (name extracted from YAML metadata). */
export function uploadBundle(yamlContent: string) {
  return request<BundleResponse>("/bundles", {
    method: "POST",
    body: JSON.stringify({ yaml_content: yamlContent }),
  })
}

/** Deploy a bundle version to an environment. */
export function deployBundle(name: string, version: number, env: string) {
  return request<DeploymentResponse>(
    `/bundles/${encodeURIComponent(name)}/${version}/deploy`,
    { method: "POST", body: JSON.stringify({ env }) },
  )
}

/** Get raw YAML for a specific bundle version. */
export async function getBundleYaml(
  name: string,
  version: number,
): Promise<string> {
  const res = await fetch(
    `${API_BASE}/bundles/${encodeURIComponent(name)}/${version}/yaml`,
    { credentials: "include" },
  )
  if (!res.ok) {
    const body = await res.text()
    throw new ApiError(res.status, body)
  }
  return res.text()
}

/** Get currently deployed bundle for a (name, env). */
export function getCurrentBundle(name: string, env: string) {
  return request<BundleResponse>(
    `/bundles/${encodeURIComponent(name)}/current?env=${encodeURIComponent(env)}`,
  )
}

// --- Evaluate (playground) ---

export interface EvaluateRequest {
  yaml_content: string
  tool_name: string
  tool_args: Record<string, unknown>
  environment?: string
  agent_id?: string
  principal?: {
    user_id?: string
    role?: string
    claims?: Record<string, unknown>
  }
}

export interface ContractEvaluation {
  id: string
  type: "pre" | "post" | "session" | "sandbox"
  matched: boolean
  effect: "deny" | "warn" | "approve" | "redact" | null
  message: string | null
}

export interface EvaluateResponse {
  verdict: string
  mode: string
  contracts_evaluated: ContractEvaluation[]
  deciding_contract: string | null
  policy_version: string
  evaluation_time_ms: number
}

export function evaluateBundle(body: EvaluateRequest) {
  return request<EvaluateResponse>("/bundles/evaluate", {
    method: "POST",
    body: JSON.stringify(body),
  })
}

// --- Deployments ---

export function listDeployments(
  bundleName?: string,
  env?: string,
  limit = 50,
) {
  const params = new URLSearchParams()
  if (bundleName) params.set("bundle_name", bundleName)
  if (env) params.set("env", env)
  params.set("limit", String(limit))
  return request<DeploymentResponse[]>(`/deployments?${params}`)
}

File: src/edictum_server/routes/bundles.py
"""Bundle CRUD and deployment endpoints — name-scoped."""

from __future__ import annotations

import base64

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from edictum_server.auth.dependencies import (
    AuthContext,
    get_current_tenant,
    require_dashboard_auth,
)
from edictum_server.config import Settings, get_settings
from edictum_server.db.engine import get_db
from edictum_server.db.models import Bundle
from edictum_server.push.manager import PushManager, get_push_manager
from edictum_server.schemas.bundles import (
    BundleCurrentResponse,
    BundleResponse,
    BundleSummaryResponse,
    BundleUploadRequest,
    BundleWithDeploymentsResponse,
    DeploymentResponse,
    DeployRequest,
)
from edictum_server.services.bundle_query_service import (
    get_bundle_enrichment,
    get_deployed_envs_by_bundle_name,
    get_deployed_envs_map,
    list_bundle_names,
)
from edictum_server.services.bundle_service import (
    get_bundle_by_version,
    get_current_bundle,
    list_bundle_versions,
    upload_bundle,
)
from edictum_server.services.deployment_service import deploy_bundle

router = APIRouter(prefix="/api/v1/bundles", tags=["bundles"])


def _bundle_to_response(bundle: Bundle) -> BundleResponse:
    """Convert a Bundle ORM instance to a BundleResponse."""
    return BundleResponse(
        id=bundle.id,
        tenant_id=bundle.tenant_id,
        name=bundle.name,
        version=bundle.version,
        revision_hash=bundle.revision_hash,
        signature_hex=bundle.signature.hex() if bundle.signature is not None else None,
        source_hub_slug=bundle.source_hub_slug,
        source_hub_revision=bundle.source_hub_revision,
        uploaded_by=bundle.uploaded_by,
        created_at=bundle.created_at,
    )


@router.post("", response_model=BundleResponse, status_code=201)
async def upload(
    body: BundleUploadRequest,
    auth: AuthContext = Depends(require_dashboard_auth),
    db: AsyncSession = Depends(get_db),
    push: PushManager = Depends(get_push_manager),
) -> BundleResponse:
    """Upload a new contract bundle (dashboard-authenticated users)."""
    try:
        bundle = await upload_bundle(
            db=db,
            tenant_id=auth.tenant_id,
            yaml_content=body.yaml_content.encode("utf-8"),
            uploaded_by=auth.email or auth.user_id or "unknown",
            source_hub_slug=None,
            source_hub_revision=None,
        )
        await db.commit()
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    push.push_to_dashboard(
        auth.tenant_id,
        {
            "type": "bundle_uploaded",
            "bundle_name": bundle.name,
            "version": bundle.version,
            "revision_hash": bundle.revision_hash,
            "uploaded_by": auth.email or auth.user_id or "unknown",
        },
    )

    return _bundle_to_response(bundle)


@router.get("", response_model=list[BundleSummaryResponse])
async def list_bundles(
    auth: AuthContext = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db),
) -> list[BundleSummaryResponse]:
    """List distinct bundle names with summaries.

    Accessible by both dashboard users and API key agents (for gate init).
    When accessed via API key, only bundles deployed to the key's environment
    are returned.
    """
    env_filter = auth.env if (auth.auth_type == "api_key" and auth.env) else None
    names = await list_bundle_names(db, auth.tenant_id, env=env_filter)
    envs_by_name = await get_deployed_envs_by_bundle_name(db, auth.tenant_id)
    enrichment = await get_bundle_enrichment(db, auth.tenant_id, env=env_filter)
    result: list[BundleSummaryResponse] = []
    for entry in names:
        bname = str(entry["name"])
        deployed_envs = envs_by_name.get(bname, [])
        # API key auth: narrow visible envs to the key's env
        if env_filter:
            deployed_envs = [env_filter] if env_filter in deployed_envs else []
        enrich = enrichment.get(bname, {})
        result.append(
            BundleSummaryResponse(
                name=bname,
                latest_version=entry["latest_version"],  # type: ignore[arg-type]
                version_count=entry["version_count"],  # type: ignore[arg-type]
                last_updated=entry["last_updated"],  # type: ignore[arg-type]
                deployed_envs=deployed_envs,
                contract_count=enrich.get("contract_count"),  # type: ignore[arg-type]
                last_deployed_at=enrich.get("last_deployed_at"),  # type: ignore[arg-type]
            )
        )
    return result


# Register /{name}/current BEFORE /{name}/{version} — FastAPI matches top-to-bottom.
# Since version is typed as int, "current" won't match it, but order is a safety measure.
@router.get("/{name}/current", response_model=BundleCurrentResponse)
async def current(
    name: str,
    env: str = Query(..., description="Target environment"),
    auth: AuthContext = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db),
) -> BundleCurrentResponse:
    """Get the currently deployed bundle for a (name, env) pair.

    Returns bundle metadata plus the YAML content (base64-encoded) so
    agents can parse and enforce contracts. Accessible by both API-key
    agents and dashboard-authenticated users.

    API key auth: the ``env`` query param must match the key's scope.
    """
    # Enforce env scope for API key auth
    if auth.auth_type == "api_key" and auth.env and env != auth.env:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"API key is scoped to '{auth.env}', cannot access env '{env}'.",
        )
    bundle = await get_current_bundle(db, auth.tenant_id, env, bundle_name=name)
    if bundle is None:
        raise HTTPException(
            status_code=404,
            detail=f"No deployed bundle '{name}' for env '{env}'",
        )
    resp = _bundle_to_response(bundle)
    return BundleCurrentResponse(
        **resp.model_dump(),
        yaml_bytes=base64.b64encode(bundle.yaml_bytes).decode(),
    )


@router.get("/{name}", response_model=list[BundleWithDeploymentsResponse])
async def list_versions(
    name: str,
    auth: AuthContext = Depends(require_dashboard_auth),
    db: AsyncSession = Depends(get_db),
) -> list[BundleWithDeploymentsResponse]:
    """List all versions for a named bundle, newest first."""
    bundles = await list_bundle_versions(db, auth.tenant_id, name)
    if not bundles:
        raise HTTPException(status_code=404, detail=f"Bundle '{name}' not found")
    envs_map = await get_deployed_envs_map(db, auth.tenant_id, name)
    return [
        BundleWithDeploymentsResponse(
            **_bundle_to_response(b).model_dump(),
            deployed_envs=envs_map.get(b.version, []),
        )
        for b in bundles
    ]


@router.get("/{name}/{version}", response_model=BundleResponse)
async def get_version(
    name: str,
    version: int,
    auth: AuthContext = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db),
) -> BundleResponse:
    """Get a specific bundle version.

    API key auth: only returns bundles that are deployed to the key's env.
    """
    bundle = await get_bundle_by_version(db, auth.tenant_id, version, bundle_name=name)
    if bundle is None:
        raise HTTPException(status_code=404, detail=f"Bundle '{name}' v{version} not found")
    # API key auth: verify this version is deployed to the key's env.
    # Return 404 (not 403) to avoid leaking version existence across envs.
    if auth.auth_type == "api_key" and auth.env:
        envs_map = await get_deployed_envs_map(db, auth.tenant_id, bundle_name=name)
        deployed_envs = envs_map.get(version, [])
        if auth.env not in deployed_envs:
            raise HTTPException(
                status_code=404,
                detail=f"Bundle '{name}' v{version} not found",
            )
    return _bundle_to_response(bundle)


@router.get("/{name}/{version}/yaml")
async def get_yaml(
    name: str,
    version: int,
    auth: AuthContext = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db),
) -> Response:
    """Get the raw YAML content of a bundle version.

    API key auth: only returns bundles that are deployed to the key's env.
    """
    bundle = await get_bundle_by_version(db, auth.tenant_id, version, bundle_name=name)
    if bundle is None:
        raise HTTPException(status_code=404, detail=f"Bundle '{name}' v{version} not found")
    # API key auth: verify this version is deployed to the key's env.
    # Return 404 (not 403) to avoid leaking version existence across envs.
    if auth.auth_type == "api_key" and auth.env:
        envs_map = await get_deployed_envs_map(db, auth.tenant_id, bundle_name=name)
        deployed_envs = envs_map.get(version, [])
        if auth.env not in deployed_envs:
            raise HTTPException(
                status_code=404,
                detail=f"Bundle '{name}' v{version} not found",
            )
    return Response(
        content=bundle.yaml_bytes,
        media_type="application/x-yaml",
    )


@router.post(
    "/{name}/{version}/deploy",
    response_model=DeploymentResponse,
    status_code=201,
)
async def deploy(
    name: str,
    version: int,
    body: DeployRequest,
    auth: AuthContext = Depends(require_dashboard_auth),
    db: AsyncSession = Depends(get_db),
    push: PushManager = Depends(get_push_manager),
    settings: Settings = Depends(get_settings),
) -> DeploymentResponse:
    """Deploy a bundle version to an environment (dashboard-authenticated)."""
    try:
        signing_secret = settings.get_signing_secret()
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    try:
        deployment = await deploy_bundle(
            db=db,
            tenant_id=auth.tenant_id,
            bundle_name=name,
            version=version,
            env=body.env,
            deployed_by=auth.email or auth.user_id or "unknown",
            signing_secret=signing_secret,
            push_manager=push,
        )
        await db.commit()
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    return DeploymentResponse(
        id=deployment.id,
        env=deployment.env,
        bundle_name=deployment.bundle_name,
        bundle_version=deployment.bundle_version,
        deployed_by=deployment.deployed_by,
        created_at=deployment.created_at,
    )
